content/things/rust-to-webassembly.md

```diff
@@ -0,0 +1,637 @@
+---json
+{
+  "title": "Rust to WebAssembly the hard way",
+  "date": "2023-02-08",
+  "socialmediaimage": "social.jpg",
+  "live": false
+}
+
+---
+
+For many, Rust is the first choice when targeting WebAssembly. Let’s take full control.
+
+<!-- more -->
+
+A long old time ago, I wrote a blog post on [how to compile C to WebAssembly without Emscripten][c to wasm], i.e. without the default tool that makes that process easy. In Rust, the tool that makes WebAssembly easy is called [wasm-bindgen], and we are going to ditch it! At the same time, Rust is a bit different in that WebAssembly was always a first-class target and the standard library is laid out to support it out of the box.
+
+## Rust to WebAssembly 101
+
+Let’s see how we can get Rust to emit WebAssembly with as little deviation from the standard Rust workflow as possible. If you look around the internet, a lot of articles and guides tell you to create a Rust library project with `cargo init --lib` and add this line to your `Cargo.toml`:
+
+|||codediff|toml
+  [package]
+  name = "my_project"
+  version = "0.1.0"
+  edition = "2021"
+  
++ [lib]
++ crate-type = ["cdylib"]
+     
+  [dependencies]
+|||
+
+Without setting the crate type to `cdylib`, the Rust compiler would emit a `.rlib` file, which is Rust’s own library format. While `cdylib` implies a dynamic library that is C-compatible, I suspect it really just stands for “use the interoperable format”, or something to that effect.
+
+For now, we'll work with the default function that Cargo generates when creating a new library:
+
+```rust
+pub fn add(left: usize, right: usize) -> usize {
+    left + right
+}
+```
+
+With all that in place, we can now compile this library to WebAssembly:
+
+```
+$ cargo run --target=wasm32-unknown-unknow --release
+```
+
+You'll find a freshly generated WebAssembly module in `target/wasm32-unknown-unknown/release/my_project.wasm`. I'll continue to use `--release` builds throughout this article as it makes the WebAssembly module a lot more readable if we want to disassemble it.
+
+### Alternative: As a binary
+
+If you are like me and adding that line to your `Cargo.toml` makes you feel weird, there’s a way around that! If your crate is designated as a binary (i.e. created via `cargo init --bin` and/or has a `main.rs` instead of a `lib.rs`), compilation to WebAssembly will succed straight away. Well, until you realize there is a lot more code in that WebAssembly module than you bargained for. You can remove the `main()` function, buth the compiler will kick off at you. You have to actively let the compiler know that the absence of `main()` is intentional by adding the `#![no_main]` crate macro to your `main.rs`. Now the emitted WebAssembly binary looks reasonable, albeit devoid of any functions.
```

Comment:
```suggestion
If you are like me and adding that line to your `Cargo.toml` makes you feel weird, there’s a way around that! If your crate is designated as a binary (i.e. created via `cargo init --bin` and/or has a `main.rs` instead of a `lib.rs`), compilation to WebAssembly will succeed straight away. Well, until you realize there is a lot more code in that WebAssembly module than you bargained for. You can remove the `main()` function, but the compiler will kick off at you. You have to actively let the compiler know that the absence of `main()` is intentional by adding the `#![no_main]` crate macro to your `main.rs.` Now the emitted WebAssembly binary looks reasonable, albeit devoid of any functions.
```